### PR TITLE
Dyno: Implement class casting primitives like `to nilable class`

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR adds support for several class-casting primitives. Specifically, it adds:

* `to nilable class` and its "checked" variant `to nilable class from type`
* `to borrowed class`+`to unmanaged class`, and their `... from unknown` checked variants

While there, I observed that we don't need to carry over the various `dtBorrowedNilable` classes from Chapel. This is because as of https://github.com/chapel-lang/chapel/pull/22501 we have a `AnyClassType` that can sit inside a decorated class type, thereby representing "any _ type", including "any borrowed nilable type". The types are necessary in the production compiler because it translates `borrowed class` into `dtBorrowedNonnilable` (e.g.) instead of representing it the way dyno does. Since these class-like builtins require special handling for "to nilable" conversions, and since they are unnecessary in Dyno, I removed them altogether.

Another thing I noticed was that we don't verify generic instantiations result in "passable" concrete formal types. In particular, we currently allow instantiating `ref x: Parent` with `shared Child`, but don't check that the resulting `shared Parent <- shared Child` conversion is allowed (it's not, due to `ref`'s semantics). To be able to properly lock down the distinction between `shared` (any shared type) and `shared class` (non-nilable shared types), I had to make sure conversions are properly caught when instantiations take place. I therefore introduce an additional `canPass` check on the instantiated formal/actual pair.

A final tweak I needed to make testing work as expected (and to improve parity with the production compiler) was to allow generic types to be passed to primitives. This was because `owned class` (e.g.) is a generic type, which Dyno's resolver uses as justification to skip resolving the call it's given to. I adjusted the resolver's logic to allow generic types as arguments to primitives. To help support this use case, I switched the 'bool' for skipping to a 'reason' enum for skipping, and spliced some of the case-by-case documentation with the enum's elements.

Reviewed by @benharsh -- thanks!

## Testing
- [x] paratest

### CODE REVIEW

File: frontend/lib/resolution/prims.cpp [hunk 1]
```diff
@@ -273,6 +274,142 @@ static QualifiedType computeDomainType(Context* context, const CallInfo& ci) {
...
+                                        const PrimCall* call,
+                                        const CallInfo& ci,
+                                        bool checked) {
+  if (ci.numActuals() != 1) return QualifiedType();
+
+  auto& actualType = ci.actual(0).type();
+  auto typePtr = actualType.type();
+  if (!typePtr) return QualifiedType();
+
+  // The cheked primitives only allow a 'type' as the actual.
```
Comment (benharsh) [inline frontend/lib/resolution/prims.cpp:376]: typo: checked

File: frontend/lib/resolution/resolution-queries.cpp [hunk 2]
```diff
@@ -1559,6 +1534,15 @@ const TypedFnSignature* instantiateSignature(Context* context,
           useType = getInstantiationType(context,
                                          actualType,
                                          formalType);
+
+          // Verify that the 'instantiation type' still accepts the actual.
+          // This might not be the case based on legal argument mapping rules.
+          auto kind = resolveIntent(useType, /* isThis */ false, /* isInit */ false);
+          auto useTypeConcrete = QualifiedType(kind, useType.type(), useType.param());
+
+          if (!canPass(context, actualType, useTypeConcrete).passes()) {
```
Comment (benharsh) [inline frontend/lib/resolution/resolution-queries.cpp:1543]: Why is this necessary?

File: frontend/test/resolution/testClassTypeclass.cpp [hunk 3]
```diff
@@ -140,9 +140,9 @@ static void test1() {
 // If the class is nilable, everything should pass.
```
Comment (benharsh) [inline frontend/test/resolution/testClassTypeclass.cpp:140]: This comment should be updated

File: frontend/lib/resolution/resolution-queries.cpp (hunk 2 above)
Comment (DanilaFe) [reply frontend/lib/resolution/resolution-queries.cpp:1543]: See the PR description. In particular, we don't validate instantiations that also require conversions for certain intents. E.g., if instantiating `ref x: Parent`, we allow it to be instantiated with `shared Child`, but `ref x: shared Parent` does not accept `shared Child`, since the `ref` intent requires the types to be equal.

File: frontend/test/resolution/testClassTypeclass.cpp [hunk 4]
```diff
@@ -140,9 +140,9 @@ static void test1() {
...
 static void test2() {
   testProgram("class?", {
-    { nonNil("owned"), shouldPass },
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply frontend/test/resolution/testClassTypeclass.cpp:143]: I've added copies of the tests with the default intents.

**Review by benharsh — APPROVED** [review]
